File: PID.cpp
#include "PID.hpp"

namespace pid_controller
{
    PID::PID(double kp, double kd, double ki, double time_interval)
    {
        this -> KP = kp;
        this -> KD = kd;
        this -> KI = ki;
        this -> integral_error = 0;
    }

    double PID::compute_adjustment(double current_error)
    {
        integral_error += time_interval * current_error;
        derivation_error = (current_error - prev_error) / time_interval;
        prev_error = current_error;

        return (KP * current_error) + (KI * integral_error) + (KD * derivation_error);
    }

    double PID::get_error() 
    {
        return prev_error;
    }
}

File: bot_setup/ros2_ws/src/bot_video/bot_video/PID.py

class PID:
    """
    PID controller.

    Attributes:
        kp (float):                 Proportional gain
        ki (float):                 Integrational gain
        kd (float):                 Derivational gain
        time_interval (float):      Time between cal. control variable 
                                    (corrective action) 
        desired_val (float):        Goal value of system 
        prev_error (float)
        integral_error (float)
        bias
    """

    def __init__(self, time_interval: float, goal: float, kp: float = 0.0, ki: 
                 float = 0.0, kd: float = 0.0, bias: float = 0.0): 
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.time_interval = time_interval
        self.integral_error = 0.0
        self.desired_val = goal
        self.prev_error = 0.0
        self.bias = bias 

    def compute_adjustment(self, current_val: float) -> float:
        error = self.desired_val - current_val 
        self.integral_error += self.time_interval * error
        derivation_error =  (error - self.prev_error) / self.time_interval 
        self.prev_error = error

        return (self.kp * error) + (self.ki * self.integral_error) + \
                (self.kd * derivation_error) + self.bias

    def getError(self):
        return self.prev_error